docs/architecture/concentrated-liquidity-adr.md

```diff
@@ -238,19 +261,292 @@ See the next `"Swap Router Module"` section of this document for more details.
+	tokenIn sdk.Coin,
+	tokenOutMinAmount sdk.Int) (tokenOutAmount sdk.Int, err error) {
+}
+```
+
+The bulk of its implementation is ported from `gamm`'s `MultihopSwapExactAmountIn`.
+Essentially, the method iterates over the routes and calls a `SwapExactAmountIn` method
+for each, subsequently updating the inter-pool swap state.
+
+The routing works by querying the index `SwapModuleRouterPrefix`,
+searching up the `swaprouterkeeper.router` mapping, and callig
+the appropriate `SwapExactAmountIn` method.
+
+```go
+// x/swaprouter/router.go RouteExactAmountIn(...)
+
+moduleRouteBytes := osmoutils.MustGet(swaproutertypes.FormatModuleRouteIndex(poolId))
+moduleRoute, _ := swaproutertypes.ModuleRouteFromBytes(moduleRouteBytes)
+
+swapModule := k.routes[moduleRoute.PoolType]
+
+_ := swapModule.SwapExactAmountIn(...)
+```
+- note that error checks and other details are omitted for brevity.
+
+Similar to pool creation logic, we are able to call `SwapExactAmountIn` on any of the swap
+modules by implementing the `SwapI` interface:
+
+```go
+// x/swaprouter/types/routes.go SwapI interface
+
+type SwapI interface {
+    ...
+
+	SwapExactAmountIn(
+		ctx sdk.Context,
+		sender sdk.AccAddress,
+		poolId gammtypes.PoolI,
+		tokenIn sdk.Coin,
+		tokenOutDenom string,
+		tokenOutMinAmount sdk.Int,
+		swapFee sdk.Dec,
+	) (sdk.Int, error)
+}
+```
+
+##### GAMM Migrations
+
+Previously we managed and stored "next pool id" and "pool creation fee" in gamm. Now, these values
+are stored in the `swaprouter` module. As a result, we perform store migration in the
+upgrade handler.
+
+Some of the queries such as `x/gamm` `NumPools depended on the "next pool id" being present in `x/gamm`.
+Since it is now moved, we introduce a new "pool count" index in `x/gamm` to keep track of the number
+of pools. TODO: do we even need this? Consider removing before release. Path forward TBD.
+
+In summary, we perform the following store migrations in the upgrade handler:
+- migrate "next pool id` from `x/gamm` to `x/swaprouter`
+- migrate "pool creation fee" from `x/gamm` to `x/swaprouter`
+- create "pool count" index in `x/gamm` TODO: do we even need this? Consider removing before release. Path forward TBD.
+
+#### GAMM Refactor
+
+> As an engineer, I would like the gamm module to be cohesive and only focus on the logic
+related to the `TraditionalAmmInterface` pool implementations.
+
+TODO: describe and document all the changes in the gamm module in more detail.
+- refer to previous sections ("Swap Router Module" and "Concentrated Liquidity Module")
+to avoid repetition.
+
+##### Swaps
+
+We rely on the pre-existing swap methods located in `x/gamm/keeper/pool.go`:
+- `SwapExactAmountIn`
+- `SwapExactAmountOut`
+
+Similarly to `concentrated-liquidity` module, these methods now implement the `swaprouter` `SwapI` interface.
+However, the concrete implementations of the methods are unchanged from before the refactor.
+
+##### New Functionality
+
+##### `InitializePool` Keeper Method
```

Comment:
We can come back to the naming right? (Just curious why we dont call it CreatePool)

Reply:
The fact of "creation" happens in `swaprouter`. That's where we call `msg.CreatePool`.

When we propagate the pool to `gamm` or `concentrated-liquidity`, it already exists. So "initialize" abstraction is more applicable then IMO